pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +380,565 @@ pub mod pallet {
+				ForeignRedemptionInfo::<T>::mutate(&who, investment_id, |info| {
+					let info = info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+
+					info.base.ensure_same_foreign(payout_foreign_currency)?;
+					info.total_tranche_tokens
+						.ensure_sub_assign(tranche_tokens_amount)?;
+
+					Ok::<_, DispatchError>(info.remaining_tranche_tokens()?)
+				})?;
+
+			T::Investment::update_redemption(
+				who,
+				investment_id,
+				T::Investment::redemption(who, investment_id)?.ensure_sub(tranche_tokens_amount)?,
+			)?;
+
+			let remaining_amount = T::Investment::redemption(who, investment_id)?;
+
+			Ok((remaining_tranche_tokens, remaining_amount))
+		}
+
+		fn collect_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			_foreign_payment_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::collect_investment(who.clone(), investment_id)
+		}
+
+		fn collect_foreign_redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			_foreign_payout_currency: T::CurrencyId,
+			_pool_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::collect_redemption(who.clone(), investment_id)
+		}
+
+		fn investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::investment(who, investment_id)
+		}
+
+		fn redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::redemption(who, investment_id)
+		}
+
+		fn accepted_payment_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapidToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let remaining_swap_amount_in = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					Pallet::<T>::unregister_swap(&who, investment_id, swap_id);
+					T::Balance::zero()
+				}
+			};
```

Comment:
IMO, we should add this information to the `Status` such that we reduce the DB reads by one.

Reply:
I keep this in mind for a future polish over these changes once we get this working. Anyway, I think the value is already cached because Orderbook has used it just before calling `notify_status_change()`. 

But I'll give thought